### PR TITLE
Vodka hangover

### LINKED ISSUE #643: Extract isCapsuleFragCorrect from MiningAdjudicator contract to a new contract
Extract `isCapsuleFragCorrect` and other pure function to new solidity contract-library (like Numerology) to split logic for CFrag correctness and other Adjudicator things

### LINKED ISSUE #749: `doubleJacobian` seems dubious
as per [this database](http://www.hyperelliptic.org/EFD/g1p/auto-shortw-jacobian.html#doubling-dbl-1998-cmo-2) and [this wiki](https://en.wikibooks.org/wiki/Cryptography/Prime_Curve/Jacobian_Coordinates#Point_Doubling_(4M_+_6S_or_4M_+_4S)), you are missing an `a * Z^4` in your calculation for `s`.

rather than using these massive constants that are hard to check, why not just use `mulmod(p - 2, s, p)`. granted, it is probably much more difficult to calculate `-8 Y^4` in this fashion. instead of calculating `2Y`, `4Y^2`, `4 X Y^2`, `-8 X Y^2`, `16 Y^4`, and then `-8 Y^4`, why not just calculate `Y^2`, `Y^4`, `-8 X`, `-8 X Y^2`, and `-8 Y^4`? this saves a `mulmod` in total i believe. this would also simplify your code with just using `p - 8` instead of massive constants.

### PR DESCRIPTION
**Pull request by cygnusv**
Cleaning up the mess after partying hard with vodka.

image

This PR just prepares a clean environment for us to work on finishing the integration of the slashing protocol with the actual re-encryption process. So far, slashing only works for mocked re-encryption. This PR addresses:
* Introduces the concept of `Task` inside a `WorkOrder`, to handle Capsules, their associated signatures, and the CFrag. This is necessary in case of something funky is detected. This PR also solidifies the protocol of signatures needed to commit Bob and Ursula to an individual re-encryption request and response, creating fundamental evidence for slashing. See above for a more formal description and #259 for some context.
* Moves re-encryption validation logic to its own contract, `ReEncryptionValidator`. (Fixes #643). Now it can be used independently from slashing (e.g., for proof of valid Ursula node)
* Removed Numerology and other unused logic. Clarifies some of the ECC-related constants. Fixes #749. 
* Improvement of MiningAdjudicator tests. (see #626, but more work required to close it)

--- 

### Building the re-encryption metadata 

In order to commit both Bob and Ursula to each re-encryption request and response, we define the following format of "_specifications_" and "_metadata_". The specification is a formal description of the request, and it's sequentially signed by Bob and Ursula. The metadata is the result of these signatures, and it's included as an integral part of the re-encryption process (that's a feature of Umbral). The resulting CFrag, which contains the metadata, is signed by Ursula. 

A more formal description follows. Let `S_b(M)` be the signature of `M` by Bob, and `S_u(M)` the signature by Ursula. Then:
```
specification = capsule || ursula_pubkey || alices_address || blockhash
task_signature = S_b(specification)
reenc_metadata = S_u(task_signature)

```
For each individual re-encryption request and response, the following dataflow applies:
```
 Bob			     Ursula
  |				|
  |				|
  |				|
  |  capsule, task_signature	|
  |---------------------------->|
  |				| reenc_metadata = S_u(task_signature)
  |				| cfrag = reencrypt(capsule, kfrag, metadata=reenc_metadata)
  |     cfrag, S_u(cfrag)	|
  |<----------------------------|
  |				|
  |				|
```
See #259 for the history of this issue and the current specification.

### CODE REVIEW

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py [hunk 1]
```diff
@@ -92,28 +94,42 @@ def fragments(metadata):
...
     # TODO: Move this to an integration test?
     umbral_params = default_params()
     u_xcoord, u_ycoord = umbral_params.u.to_affine()
     u_sign = 2 + (u_ycoord % 2)
-    assert u_sign == adjudicator_contract.functions.UMBRAL_PARAMETER_U_SIGN().call()
-    assert u_xcoord == adjudicator_contract.functions.UMBRAL_PARAMETER_U_XCOORD().call()
-    assert u_ycoord == adjudicator_contract.functions.UMBRAL_PARAMETER_U_YCOORD().call()
+    # assert u_sign == adjudicator_contract.functions.UMBRAL_PARAMETER_U_SIGN().call()
+    # assert u_xcoord == adjudicator_contract.functions.UMBRAL_PARAMETER_U_XCOORD().call()
+    # assert u_ycoord == adjudicator_contract.functions.UMBRAL_PARAMETER_U_YCOORD().call()
```
Comment (szotov) [inline tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: Since this library is separate it makes sense to make test-wrapper for library and transfer some of the tests to lib folder

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py [hunk 2]
```diff
@@ -324,70 +492,6 @@ def test_evaluate_cfrag(testerchain, escrow, adjudicator_contract):
...
-    value = escrow.functions.minerInfo(miner).call()[0]
-    reward = escrow.functions.rewardInfo(investigator).call()
-    assert not adjudicator_contract.functions.evaluatedCFrags(data_hash).call()
-    tx = adjudicator_contract.functions.evaluateCFrag(*args).transact({'from': investigator})
-    testerchain.wait_for_receipt(tx)
-    assert adjudicator_contract.functions.evaluatedCFrags(data_hash).call()
-    # Penalty was increased because it's the second violation
-    assert value - (BASE_PENALTY + PENALTY_HISTORY_COEFFICIENT) == escrow.functions.minerInfo(miner).call()[0]
-    assert reward + (BASE_PENALTY + PENALTY_HISTORY_COEFFICIENT) / REWARD_COEFFICIENT == \
-           escrow.functions.rewardInfo(investigator).call()
```
[the remark this answers is not in the document]
Comment (cygnusv) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: Yes, the second penalty is on line 323.

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py (hunk 1 above)
Comment (cygnusv) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: Agreed. I'll make the changes.

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (szotov) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: I mean slash in tests so many times that penalty will be calculated as `stake // PERCENTAGE_PENALTY_COEFFICIENT`
I think 4th violation will lead to this case

[the remark this answers is not in the document]
Comment (cygnusv) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: It happens for 3rd violation. There's already an assertion for that:
https://github.com/nucypher/nucypher/pull/863/files/ef6168c277c218484df12e5ebb22602572ac142d#diff-9c881ade6a8c6743814065c5614e27d0R419

[the remark this answers is not in the document]
Comment (szotov) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: I missed that, my bad

File: nucypher/policy/models.py [hunk 3]
```diff
@@ -534,140 +534,166 @@ def __len__(self):
...
-        else:
-            self.completed = maya.now()
-            return good_cfrags
+                raise InvalidSignature(f"{cfrag} is not properly signed by Ursula.")
+                # TODO: Instead of raising, we should do something
+
+        for task, (cfrag, reencryption_signature) in zip(self.tasks, cfrags_and_signatures):
+            task.attach_work_result(cfrag, reencryption_signature)
+
+        self.completed = maya.now()
```
Comment (tuxxy) [inline nucypher/policy/models.py]: What's the reason for keeping the timestamps?

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py [hunk 4]
```diff
@@ -229,19 +249,80 @@ def test_evaluate_cfrag(testerchain, escrow, adjudicator_contract):
...
+    # Evaluation must fail since Bob precomputed wrong values
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = adjudicator_contract.functions.evaluateCFrag(*args).transact({'from': investigator})
+        testerchain.wait_for_receipt(tx)
+
+    ###############################
+    # Test: Signatures and Metadata mismatch
+    ###############################
+
+    # TODO
```
Comment (tuxxy) [inline tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: Is this going to be in another PR?

File: nucypher/policy/models.py [hunk 5]
```diff
@@ -534,140 +534,166 @@ def __len__(self):
 
 class WorkOrder:
 
-    class NotFromBob(InvalidSignature):
-        def __init__(self):
-            super().__init__("This doesn't appear to be from Bob.")
+    class Task:
```
[the remark this answers is not in the document]
Comment (cygnusv) [reply nucypher/policy/models.py]: This is completely open to suggestions. What about `Work` or `Job`?

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py (hunk 4 above)
Comment (cygnusv) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: Yes, additional work is needed before we can test anything.

File: nucypher/policy/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/policy/models.py]: `.do_the(work)`

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/policy/models.py]: Yeah it feels like we've gone a bit too-generic with naming scheme, but I can live with it.

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py (hunk 4 above)
Comment (tuxxy) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: Can you link the newly created issues in the PR, please?

File: nucypher/policy/models.py [hunk 6]
```diff
@@ -534,140 +534,166 @@ def __len__(self):
...
-        alice_address_signature = bytes(self.alice_address_signature)
         ursula_verifying_key = self.ursula.stamp.as_umbral_pubkey()
 
-        for counter, capsule in enumerate(self.capsules):
-            cfrag, signature = cfrags_and_signatures[counter]
-
-            # Validate CFrag metadata
-            capsule_signature = bytes(self.capsule_signatures[counter])
-            metadata_input = capsule_signature + alice_address_signature
+        for task, (cfrag, reencryption_signature) in zip(self.tasks, cfrags_and_signatures):
```
Comment (KPrasch) [inline nucypher/policy/models.py]: :+1:

File: nucypher/policy/models.py [hunk 7]
```diff
@@ -534,140 +534,166 @@ def __len__(self):
...
     def construct_by_bob(cls, arrangement_id, capsules, ursula, bob):
         alice_verifying_key = capsules[0].get_correctness_keys()["verifying"]
+        alice_address = canonical_address_from_umbral_key(alice_verifying_key)
+
+        # TODO: Bob's input to prove freshness for this work order
+        blockhash = b'\x00' * 32
 
-        capsules_bytes = []
-        capsule_signatures = []
+        tasks, tasks_bytes = [], []
```
Comment (KPrasch) [inline nucypher/policy/models.py]: Am I the only one who enjoys reading `list()`?  :smile:

Comment (michwill) [reply nucypher/policy/models.py]: I think so. I like `[]` more!

File: nucypher/policy/models.py [hunk 8]
```diff
@@ -534,140 +534,166 @@ def __len__(self):
...
-            if signature.verify(bytes(cfrag) + bytes(capsule), ursula_verifying_key):
+            # Validate re-encryption signatures
+            if reencryption_signature.verify(bytes(cfrag), ursula_verifying_key):
                 good_cfrags.append(cfrag)
             else:
-                raise InvalidSignature("{} is not properly signed by Ursula.".format(cfrag))
-        else:
-            self.completed = maya.now()
-            return good_cfrags
+                raise InvalidSignature(f"{cfrag} is not properly signed by Ursula.")
```
Comment (KPrasch) [inline nucypher/policy/models.py]: I'm not exactly sure where this PR falls in the timeline of slashing integration, but I noticed that there is little test coverage on these `raise` conditions, including this one.

File: nucypher/policy/models.py (hunk 3 above)
Comment (cygnusv) [reply nucypher/policy/models.py]: You'll have to ask that to our mutual friend @jMyles 
(https://github.com/nucypher/nucypher/commit/20bb613a472c1a41445ad52744470427c21a69d7)

File: tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py (hunk 4 above)
Comment (cygnusv) [reply tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py]: #626

File: nucypher/policy/models.py (hunk 8 above)
Comment (cygnusv) [reply nucypher/policy/models.py]: #626

**Comment by cygnusv** [comment]
@tuxxy @KPrasch @szotov I broadened the scope of https://github.com/nucypher/nucypher/issues/626 to track the improvement of tests of this logic. 
It will be my next PR. Worst case, right after #899. Pinky promise.
